react/features/mobile/callkit/middleware.js

```diff
@@ -0,0 +1,194 @@
+} from '../../app';
+import {
+    CONFERENCE_FAILED,
+    CONFERENCE_LEFT,
+    CONFERENCE_WILL_JOIN,
+    CONFERENCE_JOINED
+} from '../../base/conference';
+import { getInviteURL } from '../../base/connection';
+import {
+    SET_AUDIO_MUTED,
+    SET_VIDEO_MUTED,
+    isVideoMutedByAudioOnly,
+    setAudioMuted
+} from '../../base/media';
+import { MiddlewareRegistry, toState } from '../../base/redux';
+import { _SET_CALLKIT_LISTENERS } from './actionTypes';
+import CallKit from './CallKit';
+
+/**
+ * Middleware that captures several system actions and hooks up CallKit.
+ *
+ * @param {Store} store - The redux store.
+ * @returns {Function}
+ */
+MiddlewareRegistry.register(({ dispatch, getState }) => next => action => {
+    const result = next(action);
+
+    switch (action.type) {
+    case _SET_CALLKIT_LISTENERS: {
+        const { listeners } = getState()['features/callkit'];
+
+        if (listeners) {
+            for (const [ event, listener ] of listeners) {
+                CallKit.removeEventListener(event, listener);
+            }
+        }
+
+        if (action.listeners) {
+            for (const [ event, listener ] of action.listeners) {
+                CallKit.addEventListener(event, listener);
+            }
+        }
+
+        break;
+    }
+
+    case APP_WILL_MOUNT: {
+        CallKit.setup();  // TODO: set app icon.
+        const listeners = new Map();
+        const callEndListener = data => {
+            const conference = getCurrentConference(getState);
+
+            if (conference && conference.callUUID === data.callUUID) {
+                // We arrive here when a call is ended by the system, for
+                // for example when another incoming call is received and the
+                // user selects "End & Accept".
+                delete conference.callUUID;
+                dispatch(appNavigate(undefined));
+            }
+        };
+
+        listeners.set('performEndCallAction', callEndListener);
+
+        // Set the same listener for providerDidReset. According to the docs,
+        // when the system resets we should terminate all calls.
+        listeners.set('providerDidReset', callEndListener);
+
+        const setMutedListener = data => {
+            const conference = getCurrentConference(getState);
+
+            if (conference && conference.callUUID === data.callUUID) {
+                // Break the loop. Audio can be muted both from the CallKit
+                // interface and from the Jitsi Meet interface. We must keep
+                // them in sync, but at some point the loop needs to be broken.
+                // We are doing it here, on the CallKit handler.
+                const { muted } = getState()['features/base/media'].audio;
+
+                if (muted !== data.muted) {
+                    dispatch(setAudioMuted(Boolean(data.muted)));
+                }
+
+            }
+        };
+
+        listeners.set('performSetMutedCallAction', setMutedListener);
```

Comment:
I found dealing with `Map` and `removeEventListener` too complex. `EventEmitter.addListener` returns a subscription which can be `remove`d.